docs/features/source-generators.md

```diff
@@ -0,0 +1,291 @@
+and the design of the feature will need to keep this in mind.
+
+For 1st party generators, especially Razor and Blazor, we aim at a minimum to match the existing
+performance seen by users today. It is expected that even naïve generator-based implementations 
+will perform significantly faster than the existing tooling, due to less communication overhead
+and duplicated work, but improving the speed of these experiences is not a primary goal of this project.
+
+### Language Changes
+
+This design does not currently propose altering the language, it is purely a compiler feature. 
+The previous design for source generators introduced the `replace` and `original` keywords. 
+This proposal removes these, as the source generated is purely additional and so there is no
+need for them. We expect that most scenarios are possible with the existing use of `partial` 
+definitions; as a V1 we expect to ship in this state. If concrete scenarios are later shown
+that can’t be achieved with the V1 approach we would consider allowing modification as a V2. 
+
+## Use cases
+
+We've identified several first and third party candidates that would benefit from source generators:
+
+- ASP.Net: Improve startup time 
+- Blazor and Razor: Massively reduce tooling burden 
+- Azure Functions: regex compilation during startup
+- Azure SDK
+- [gRPC](https://docs.microsoft.com/en-us/aspnet/core/grpc/?view=aspnetcore-3.1)
+- Resx file generation
+- [System.CommandLine](https://github.com/dotnet/command-line-api)
+- Serializers
+- [SWIG](http://www.swig.org/)
+
+
+## Discussion / Open Issues / TODOs:
+
+**Interface vs Class for ISourceGenerator**: 
+
+We discussed about this being an interface or class. Analyzers chose to have a abstract base class,
+but we weren't sure what we'd end up a need since ultimately we only had one method on this.
+Keeping it an interface also was more natural since we have other interfaces that 
+implement this interface as well for optional light-up.
+
+**IDependsOnCompilationGenerator**:
+
+We did discuss if there should be an IDependsOnCompilationGenerator to formally state 
+that you actually use a compilation. After all, if you don't use the compilation 
+then we know your performance in the IDE is greatly simplified. However every 
+scenario we've had for reading additional files has also needed the compilation, 
+so we simply weren't sure what that was going to bring.
+
+**Breakpoints in generated files**: 
+
+Do we map this back to the in-memory file?
+
+**Should generators be push or pull**:
+
+Source generators are pull-based, analyzers are push-based (registration based). Should
+we use a push-based model for generators as well?
+
+- If we go down the push-based model, walking the tree should make sure to continue
+to produce events for as many nodes as possible, even with errors, as generators
+will often work in the presence 
+
+- The events that we use today for analyzers may require may more work to produce,
```

Comment:
By "more work", specifically, doing the minimal amount of work needed to produce an event, i.e. creating a Compilation should not need to do any binding.